File: typescript/src/rappid-card/artifacts.ts
import { mkdir, writeFile } from 'node:fs/promises';
import { join, resolve } from 'node:path';

import {
  RAPPID_CARD_FILENAME,
} from './types.js';
import {
  RAPPID_CARD_FIXTURE_NAMES,
  buildRappidCardFixture,
  listRappidCardFixtures,
} from './fixtures.js';
import {
  renderRappidCardQrPng,
  renderRappidCardQrSvg,
} from './qr.js';

export type QrArtifactFormat = 'svg' | 'png' | 'both';

export interface FixtureDeckWriteResult {
  directory: string;
  fixtures: number;
  files: string[];
}

export async function writeRappidCardFixtureDeck(
  directory: string,
  format: QrArtifactFormat = 'svg',
): Promise<FixtureDeckWriteResult> {
  const root = resolve(directory);
  await mkdir(root, { recursive: true });
  const files: string[] = [];
  const deck = {
    schema: 'rappid-card-fixture-deck/1',
    fixtures: listRappidCardFixtures(),
  };
  const deckPath = join(root, 'deck.json');
  await writeFile(deckPath, `${JSON.stringify(deck, null, 2)}\n`, 'utf8');
  files.push(deckPath);

  for (const name of RAPPID_CARD_FIXTURE_NAMES) {
    const fixture = buildRappidCardFixture(name);
    const fixtureDirectory = join(root, name);
    await mkdir(fixtureDirectory, { recursive: true });
    const cardPath = join(fixtureDirectory, RAPPID_CARD_FILENAME);
    const linkPath = join(fixtureDirectory, 'rappid-card.link.txt');
    const policyPath = join(fixtureDirectory, 'rappid-card.policy.json');
    const authorizationPath = join(
      fixtureDirectory,
      'rappid-card.authorization.json',
    );
    const revocationsPath = join(
      fixtureDirectory,
      'rappid-card.revocations.json',
    );
    await writeFile(
      cardPath,
      `${JSON.stringify(fixture.manifest, null, 2)}\n`,
      'utf8',
    );
    await writeFile(linkPath, `${fixture.deepLink}\n`, 'utf8');
    await writeFile(
      policyPath,
      `${JSON.stringify(fixture.policy, null, 2)}\n`,
      'utf8',
    );
    await writeFile(
      authorizationPath,
      `${JSON.stringify(fixture.authorization, null, 2)}\n`,
      'utf8',
    );
    await writeFile(
      revocationsPath,
      `${JSON.stringify(fixture.revocations, null, 2)}\n`,
      'utf8',
    );
    files.push(
      cardPath,
      linkPath,
      policyPath,
      authorizationPath,
      revocationsPath,
    );
    if (format === 'svg' || format === 'both') {
      const svgPath = join(fixtureDirectory, 'rappid-card.svg');
      await writeFile(
        svgPath,
        await renderRappidCardQrSvg(fixture.deepLink),
        'utf8',
      );
      files.push(svgPath);
    }
    if (format === 'png' || format === 'both') {
      const pngPath = join(fixtureDirectory, 'rappid-card.png');
      await writeFile(
        pngPath,
        await renderRappidCardQrPng(fixture.deepLink),
      );
      files.push(pngPath);
    }
  }
  return {
    directory: root,
    fixtures: RAPPID_CARD_FIXTURE_NAMES.length,
    files,
  };
}

File: python/openrappter/rappid_card/artifacts.py
"""Deterministic on-disk fixture deck for developers."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict

from .fixtures import (
    RAPPID_CARD_FIXTURE_NAMES,
    build_rappid_card_fixture,
    list_rappid_card_fixtures,
)
from .qr import render_rappid_card_qr_png, render_rappid_card_qr_svg
from .types import RAPPID_CARD_FILENAME


def write_rappid_card_fixture_deck(
    directory: str, qr_format: str = "svg"
) -> Dict[str, Any]:
    if qr_format not in {"svg", "png", "both"}:
        raise ValueError("format must be svg, png, or both")
    root = Path(directory).resolve()
    root.mkdir(parents=True, exist_ok=True)
    files = []
    deck_path = root / "deck.json"
    deck_path.write_text(
        json.dumps(
            {
                "schema": "rappid-card-fixture-deck/1",
                "fixtures": list_rappid_card_fixtures(),
            },
            indent=2,
        )
        + "\n",
        encoding="utf-8",
    )
    files.append(str(deck_path))
    for name in RAPPID_CARD_FIXTURE_NAMES:
        fixture = build_rappid_card_fixture(name)
        fixture_directory = root / name
        fixture_directory.mkdir(parents=True, exist_ok=True)
        card_path = fixture_directory / RAPPID_CARD_FILENAME
        link_path = fixture_directory / "rappid-card.link.txt"
        card_path.write_text(
            json.dumps(fixture.manifest, indent=2) + "\n",
            encoding="utf-8",
        )
        link_path.write_text(fixture.deep_link + "\n", encoding="utf-8")
        files.extend([str(card_path), str(link_path)])
        if qr_format in {"svg", "both"}:
            svg_path = fixture_directory / "rappid-card.svg"
            svg_path.write_text(
                render_rappid_card_qr_svg(fixture.deep_link),
                encoding="utf-8",
            )
            files.append(str(svg_path))
        if qr_format in {"png", "both"}:
            png_path = fixture_directory / "rappid-card.png"
            png_path.write_bytes(render_rappid_card_qr_png(fixture.deep_link))
            files.append(str(png_path))
    return {
        "directory": str(root),
        "fixtures": len(RAPPID_CARD_FIXTURE_NAMES),
        "files": files,
    }
